sql/src/main/java/io/crate/planner/node/ddl/OptimizeTablePlan.java

```diff
@@ -0,0 +1,174 @@
+import static io.crate.analyze.OptimizeTableSettings.FLUSH;
+import static io.crate.analyze.OptimizeTableSettings.MAX_NUM_SEGMENTS;
+import static io.crate.analyze.OptimizeTableSettings.ONLY_EXPUNGE_DELETES;
+import static io.crate.analyze.OptimizeTableSettings.SUPPORTED_SETTINGS;
+import static io.crate.analyze.OptimizeTableSettings.UPGRADE_SEGMENTS;
+import static io.crate.analyze.PartitionPropertiesAnalyzer.toPartitionName;
+import static io.crate.data.SentinelRow.SENTINEL;
+
+public class OptimizeTablePlan implements Plan {
+
+    private final AnalyzedOptimizeTable optimizeTable;
+
+    public OptimizeTablePlan(AnalyzedOptimizeTable optimizeTable) {
+        this.optimizeTable = optimizeTable;
+    }
+
+    @Override
+    public StatementType type() {
+        return StatementType.DDL;
+    }
+
+    @Override
+    public void executeOrFail(DependencyCarrier dependencies,
+                              PlannerContext plannerContext,
+                              RowConsumer consumer,
+                              Row parameters,
+                              SubQueryResults subQueryResults) {
+        if (optimizeTable.tables().isEmpty()) {
+            consumer.accept(InMemoryBatchIterator.empty(SENTINEL), null);
+            return;
+        }
+
+        OptimizeTableAnalyzedStatement stmt = createStatement(
+            optimizeTable,
+            plannerContext.transactionContext(),
+            plannerContext.functions(),
+            parameters,
+            subQueryResults
+        );
+
+        var settings = stmt.settings();
+        var toOptimize = stmt.tablesToOptimize();
+
+        if (UPGRADE_SEGMENTS.get(settings)) {
+            var request = new UpgradeRequest(toOptimize.toArray(new String[0]));
+
+            var transportUpgradeAction = dependencies.transportActionProvider().transportUpgradeAction();
+            transportUpgradeAction.execute(request, new OneRowActionListener<>(
+                consumer,
+                response -> new Row1(toOptimize.isEmpty() ? -1L : (long) toOptimize.size()))
+            );
+        } else {
+            var request = new ForceMergeRequest(toOptimize.toArray(new String[0]));
+            request.maxNumSegments(MAX_NUM_SEGMENTS.get(settings));
+            request.onlyExpungeDeletes(ONLY_EXPUNGE_DELETES.get(settings));
+            request.flush(FLUSH.get(settings));
+            request.indicesOptions(IndicesOptions.lenientExpandOpen());
+
+            var transportForceMergeAction = dependencies.transportActionProvider().transportForceMergeAction();
+            transportForceMergeAction.execute(request, new OneRowActionListener<>(
+                consumer,
+                response -> new Row1(toOptimize.isEmpty() ? -1L : (long) toOptimize.size()))
+            );
+        }
+    }
+
+    @VisibleForTesting
+    public static OptimizeTableAnalyzedStatement createStatement(AnalyzedOptimizeTable optimizeTable,
```

Comment:
I think it would be nicer if we removed the intermediate `OptimizeTableAnalyzedStatement` and instead directly build the request classes. WDYT?

Reply:
It was my intention but then i ended up with `OptimizeTableAnalyzedStatement` such as we have two code path depends on the setting that would result in two different type of requests. It didn't look well neither in plan class nor in tests, so i've chosen to still keep `OptimizeTableAnalyzedStatement` :/